source/secure_tunneling.c

```diff
@@ -42,6 +43,97 @@ static void s_on_websocket_shutdown(struct aws_websocket *websocket, int error_c
     UNUSED(user_data);
 }
 
+static bool s_on_websocket_incoming_frame_begin(
+    struct aws_websocket *websocket,
+    const struct aws_websocket_incoming_frame *frame,
+    void *user_data) {
+    UNUSED(websocket);
+    UNUSED(frame);
+    UNUSED(user_data);
+    return true;
+}
+
+static void s_process_iot_st_msg(struct aws_secure_tunnel *secure_tunnel, struct aws_iot_st_msg *st_msg) {
+    /* TODO: Check streamId */
+    /* secure_tunnel->stream_id != st_msg->streamId */
+
+    switch (st_msg->type) {
+        case UNKNOWN:
+            break;
+        case DATA:
+            secure_tunnel->config.on_data_receive(secure_tunnel, &st_msg->payload);
+            break;
+        case STREAM_START:
+            break;
+        case STREAM_RESET:
+            break;
+        case SESSION_RESET:
+            break;
+        default:
+            break;
+    }
+}
+
+static void s_process_received_data(struct aws_secure_tunnel *secure_tunnel) {
+    struct aws_byte_buf *received_data = &secure_tunnel->received_data;
+    struct aws_byte_cursor cursor = aws_byte_cursor_from_buf(received_data);
+
+    uint16_t data_length = 0;
+    struct aws_byte_cursor tmp_cursor =
+        cursor; // If there are at least two bytes for the data_length, but not enough
+                // data for a complete secure tunnel frame, we don't want to move `cursor`.
+    while (aws_byte_cursor_read_be16(&tmp_cursor, &data_length) && tmp_cursor.len >= data_length) {
+        cursor = tmp_cursor;
+
+        struct aws_byte_cursor st_frame = {.len = data_length, .ptr = cursor.ptr};
+        aws_byte_cursor_advance(&cursor, data_length);
+
+        struct aws_iot_st_msg st_msg;
+        aws_iot_st_msg_deserialize_from_cursor(&st_msg, &st_frame, secure_tunnel->config.allocator);
+        s_process_iot_st_msg(secure_tunnel, &st_msg);
+    }
+
+    if (cursor.ptr != received_data->buffer) {
```

Comment:
At this point, `cursor` points to an incomplete secure tunnel frame. We can't de-serialize this frame yet. We need to wait for more websocket frames to complete this last secure tunnel frame.

I am copying this incomplete frame to the beginning of the byte buffer so that when the next websocket frame comes in, we can continue to process the data from the beginning. 

I do realize there is byte copying and put a TODO here to investigate if there is a better data structure such as circular buffer. The aws-c-common project does have a ring_buffer. But since secure tunnel frame and websocket frame does not need to be on the same boundary, so I can't use ring_buffer.

For reference:
https://github.com/aws-samples/aws-iot-securetunneling-localproxy/blob/master/WebsocketProtocolGuide.md#tunneling-message-frames